src/OpenLoco/src/Paint/PaintBuilding.cpp

```diff
@@ -0,0 +1,206 @@
+        {
+            const auto* scaffObj = ObjectManager::get<ScaffoldingObject>();
+            const auto segmentHeight = scaffObj->segmentHeights[scaffSegType];
+            const auto& scaffImages = isMultiTile ? getScaffoldingImages(scaffSegType).get2x2() : getScaffoldingImages(scaffSegType).get1x1();
+            uint32_t baseScaffImageIdx = scaffObj->image;
+            ImageId baseScaffImage{};
+            if (elBuilding.isGhost())
+            {
+                baseScaffImage = Gfx::applyGhostToImage(baseScaffImageIdx);
+            }
+            else
+            {
+                baseScaffImage = ImageId(baseScaffImageIdx, scaffoldingColour);
+            }
+            auto scaffImage = baseScaffImage.withIndexOffset(scaffImages.front);
+            auto segmentImageOffset = imageOffset;
+            for (int8_t remainingHeight = totalSectionHeight; remainingHeight > 0; remainingHeight -= segmentHeight, segmentImageOffset.z += segmentHeight)
+            {
+                session.addToPlotListAsChild(scaffImage, segmentImageOffset, bbOffset, bbSize);
+            }
+            scaffImage = baseScaffImage.withIndexOffset(scaffImages.getRoof(rotation));
+            session.addToPlotListAsChild(scaffImage, segmentImageOffset, bbOffset, bbSize);
+        }
+    }
+
+    // 0x0042C6C4
+    void paintBuilding(PaintSession& session, const World::BuildingElement& elBuilding)
+    {
+        session.setItemType(Ui::ViewportInteraction::InteractionItem::building);
+        auto* buildingObj = elBuilding.getObject();
+        if (buildingObj->hasFlags(BuildingObjectFlags::isHeadquarters))
+        {
+            session.setItemType(Ui::ViewportInteraction::InteractionItem::headquarterBuilding);
+        }
+
+        ImageId baseColour(0, elBuilding.colour());
+        if (elBuilding.isGhost())
+        {
+            baseColour = Gfx::applyGhostToImage(0);
+        }
+        // Combine this with any imageId
+        const uint8_t rotation = (session.getRotation() + elBuilding.rotation()) & 0x3;
+
+        // 0x00525D4E
+        const int16_t bbLengthZ = std::min(elBuilding.clearHeight() - elBuilding.baseHeight(), 128) - 2;
+
+        // 0x00525D18
+        const auto variation = elBuilding.variation();
+
+        const auto baseHeight = elBuilding.baseHeight();
+
+        bool isMultiTile = buildingObj->hasFlags(BuildingObjectFlags::largeTile);
+
+        // Note: Image offsets will change as you move up the building but bboffset/size does not
+        const World::Pos3 imageOffset = (isMultiTile ? kImageOffsetBase2x2 : kImageOffsetBase1x1) + World::Pos3{ 0, 0, baseHeight };
+        const World::Pos3 bbOffset = (isMultiTile ? kBBOffsetBase2x2 : kBBOffsetBase1x1) + World::Pos3{ 0, 0, baseHeight };
+        const World::Pos3 bbSize = (isMultiTile ? kBBSizeBase2x2 : kBBSizeBase1x1) + World::Pos3{ 0, 0, bbLengthZ };
+
+        session.resetLastPS(); // Odd...
+        if (buildingObj->hasFlags(BuildingObjectFlags::hasShadows))
```

Comment:
Interestingly no object seems to have hasShadows. I just know what it does from the industry object.